snmalloc-edp/tests/global_alloc.rs

```diff
@@ -0,0 +1,108 @@
+use std::{alloc::{self, GlobalAlloc}, cell::Cell, ptr};
+
+use snmalloc_edp::*;
+
+thread_local! {
+    static THREAD_ALLOC: Cell<*mut Alloc> = const { Cell::new(ptr::null_mut()) };
+}
+
+#[no_mangle]
+pub fn __rust_get_thread_allocator() -> *mut Alloc {
+    THREAD_ALLOC.get()
+}
+
+struct System;
+
+unsafe impl alloc::GlobalAlloc for System {
+    #[inline]
+    unsafe fn alloc(&self, layout: alloc::Layout) -> *mut u8 {
+        // SAFETY: the caller must uphold the safety contract for `malloc`
+        sn_rust_alloc(layout.align(), layout.size())
+    }
+
+    #[inline]
+    unsafe fn alloc_zeroed(&self, layout: alloc::Layout) -> *mut u8 {
+        // SAFETY: the caller must uphold the safety contract for `malloc`
+        sn_rust_alloc_zeroed(layout.align(), layout.size())
+    }
+
+    #[inline]
+    unsafe fn dealloc(&self, ptr: *mut u8, layout: alloc::Layout) {
+        // SAFETY: the caller must uphold the safety contract for `malloc`
+        sn_rust_dealloc(ptr, layout.align(), layout.size())
+    }
+
+    #[inline]
+    unsafe fn realloc(&self, ptr: *mut u8, layout: alloc::Layout, new_size: usize) -> *mut u8 {
+        // SAFETY: the caller must uphold the safety contract for `malloc`
+        sn_rust_realloc(ptr, layout.align(), layout.size(), new_size)
+    }
+}
+
+// SAFETY: this should only be called once per thread, and the global
+// allocator shouldn't be used outside of this function
+unsafe fn with_thread_allocator<F: FnOnce() -> R, R>(f: F) -> R {
+    unsafe {
+        let mut allocator = std::mem::MaybeUninit::<Alloc>::uninit();
+        sn_thread_init(allocator.as_mut_ptr());
+        THREAD_ALLOC.set(allocator.as_mut_ptr());
+
+        let r = f();
+
+        THREAD_ALLOC.set(ptr::null_mut());
+        sn_thread_cleanup(allocator.as_mut_ptr());
+
+        r
+    }
+}
+
+#[test]
+fn test() {
```

Comment:
I know that this isn't the only testing we have for the allocator/snmalloc, but I think we should have at least a little more testing here. At a minimum, I think we should exercise all of the alloc/alloc_zeroed/dealloc/realloc functions.

Reply:
This test also seems to be failing with a signal 11 (SIGSEGV), I need to investigate what the issue is. But when snmalloc-edp is plugged into the rust compiler's standard library, I see all the tests assembly and run-make tests (tests included in CI) succeed.